Fundamentals/bin_expo.cpp

```diff
@@ -0,0 +1,52 @@
+/*
+ * @author Samarth Agarwal
+ *
+ * Binary Exponentiation is used to find 
+ * x raised to power y modulo p in O(log y) time.
+ *
+ * Raising a to the power of n is expressed naively as multiplication by a 
+ * done n−1 times: a^n = a * a * ... * a. However, this approach is not 
+ * practical for large a or n.
+ *
+ * now a^(b + c) = (a^b) * (a^c) and a^(2b) = (a^b)^2.
+ *
+ * The idea of binary exponentiation is, that we split the work using the 
+ * binary representation of the exponent.
+ */
+
+#include <iostream>
+
+using namespace std;
+
+
+/*
+ * Time complexity: O(log(y))
+ * Space Complexity: O(1)
+ */
+
+int bin_expo(int x, int y, int p) {
+	int res = 1;
+
+	x %= p;
+
+	while(y) {
+		if(y%2)
+			res = (res * 1LL * x) % p;
```

Comment:
Type cast res instead of multiplying by 1LL, it saves one instruction.

```cpp
res = ((long long)res * x) % p;
```

and justify why is it used as it may not be apparent to beginners.